fibheap/fibheap.go

```diff
@@ -0,0 +1,392 @@
+* for single-source shortest paths can be shown to run in O(m + n lg n) using
+* a Fibonacci heap, compared to O(m lg n) using a standard binary or binomial
+* heap.
+*
+* Internally, a Fibonacci heap is represented as a circular, doubly-linked
+* list of trees obeying the min-heap property.  Each node stores pointers
+* to its parent (if any) and some arbitrary child.  Additionally, every
+* node stores its degree (the number of children it has) and whether it
+* is a "marked" node.  Finally, each Fibonacci heap stores a pointer to
+* the tree with the minimum value.
+*
+* To insert a node into a Fibonacci heap, a singleton tree is created and
+* merged into the rest of the trees.  The merge operation works by simply
+* splicing together the doubly-linked lists of the two trees, then updating
+* the min pointer to be the smaller of the minima of the two heaps.  Peeking
+* at the smallest element can therefore be accomplished by just looking at
+* the min element.  All of these operations complete in O(1) time.
+*
+* The tricky operations are dequeueMin and decreaseKey.  dequeueMin works
+* by removing the root of the tree containing the smallest element, then
+* merging its children with the topmost roots.  Then, the roots are scanned
+* and merged so that there is only one tree of each degree in the root list.
+* This works by maintaining a dynamic array of trees, each initially null,
+* pointing to the roots of trees of each dimension.  The list is then scanned
+* and this array is populated.  Whenever a conflict is discovered, the
+* appropriate trees are merged together until no more conflicts exist.  The
+* resulting trees are then put into the root list.  A clever analysis using
+* the potential method can be used to show that the amortized cost of this
+* operation is O(lg n), see "Introduction to Algorithms, Second Edition" by
+* Cormen, Rivest, Leiserson, and Stein for more details.
+*
+* The other hard operation is decreaseKey, which works as follows.  First, we
+* update the key of the node to be the new value.  If this leaves the node
+* smaller than its parent, we're done.  Otherwise, we cut the node from its
+* parent, add it as a root, and then mark its parent.  If the parent was
+* already marked, we cut that node as well, recursively mark its parent,
+* and continue this process.  This can be shown to run in O(1) amortized time
+* using yet another clever potential function.  Finally, given this function,
+* we can implement delete by decreasing a key to -\infty, then calling
+* dequeueMin to extract it.
+ */
+
+package fibheap
+
+import (
+	"fmt"
+	"math"
+)
+
+/******************************************
+ ************** INTERFACE *****************
+ ******************************************/
+
+// The FloatingFibonacciHeap is an implementation of a fibonacci heap
+// with only floating-point priorities and no user data attached.
+type FloatingFibonacciHeap interface {
```

Comment:
Why an interface instead of just directly exposing the a fib heap struct with the same public methods?